staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_test.go

```diff
@@ -501,3 +505,166 @@ func TestReflectorForWatchCache(t *testing.T) {
+			interval:           eventFreshDuration / 6,
+			expectCapacity:     10,
+		},
+		{
+			name:               "[singular capacity] events outside eventFreshDuration without change cache capacity",
+			eventCount:         5,
+			cacheCapacity:      5,
+			lowerBoundCapacity: 5 / 2,
+			upperBoundCapacity: 5 * 2,
+			interval:           eventFreshDuration / 5,
+			expectCapacity:     5,
+		},
+		{
+			name:               "[singular capacity] half recent events outside eventFreshDuration cause cache shrinking",
+			eventCount:         5,
+			cacheCapacity:      5,
+			lowerBoundCapacity: 5 / 2,
+			upperBoundCapacity: 5 * 2,
+			interval:           eventFreshDuration/2 + time.Second,
+			expectCapacity:     2,
+		},
+		{
+			name:               "[singular capacity] half recent events outside eventFreshDuration cause cache shrinking with given lowerBoundCapacity",
+			eventCount:         5,
+			cacheCapacity:      5,
+			lowerBoundCapacity: 3,
+			upperBoundCapacity: 5 * 2,
+			interval:           eventFreshDuration/2 + time.Second,
+			expectCapacity:     3,
+		},
+		{
+			name:               "[singular capacity] events inside eventFreshDuration cause cache expanding with given upperBoundCapacity",
+			eventCount:         5,
+			cacheCapacity:      5,
+			lowerBoundCapacity: 5 / 2,
+			upperBoundCapacity: 8,
+			interval:           eventFreshDuration / 6,
+			expectCapacity:     8,
+		},
+		{
+			name:               "[even capacity] events inside eventFreshDuration cause cache expanding",
+			eventCount:         6,
+			cacheCapacity:      6,
+			lowerBoundCapacity: 6 / 2,
+			upperBoundCapacity: 6 * 2,
+			interval:           eventFreshDuration / 7,
+			expectCapacity:     12,
+		},
+		{
+			name:               "[even capacity] events outside eventFreshDuration without change cache capacity",
+			eventCount:         6,
+			cacheCapacity:      6,
+			lowerBoundCapacity: 6 / 2,
+			upperBoundCapacity: 6 * 2,
+			interval:           eventFreshDuration / 6,
+			expectCapacity:     6,
+		},
+		{
+			name:               "[even capacity] half recent events outside eventFreshDuration cause cache shrinking",
+			eventCount:         6,
+			cacheCapacity:      6,
+			lowerBoundCapacity: 6 / 2,
+			upperBoundCapacity: 6 * 2,
+			interval:           eventFreshDuration / 2,
+			expectCapacity:     3,
+		},
+		{
+			name:               "[even capacity] half recent events outside eventFreshDuration cause cache shrinking with given lowerBoundCapacity",
+			eventCount:         6,
+			cacheCapacity:      6,
+			lowerBoundCapacity: 4,
+			upperBoundCapacity: 6 * 2,
+			interval:           eventFreshDuration / 2,
+			expectCapacity:     4,
+		},
+		{
+			name:               "[even capacity] events inside eventFreshDuration cause cache expanding with given upperBoundCapacity",
+			eventCount:         6,
+			cacheCapacity:      6,
+			lowerBoundCapacity: 6 / 2,
+			upperBoundCapacity: 10,
+			interval:           eventFreshDuration / 7,
+			expectCapacity:     10,
+		},
```

Comment:
The other cases that I would like to see covered are those that have startIndex != 0

Reply:
Done.